tests/framework/umbra_client_faulty_redzone.c

```diff
@@ -93,73 +97,110 @@ dr_client_main(client_id_t id, int argc, const char *argv[])
                                    1) != DRMF_SUCCESS) {
         DR_ASSERT(false);
     }
 
     /* Use a displacement of a page size to access ahead and try to hit a faulty
-     * redzone.
+     * redzone. Naturally, this does not guarantee that the red-zone will always
+     * be accessed, but we hope that it will hit at least once to trigger a fault.
      */
     instrlist_meta_preinsert(
         ilist, where,
         INSTR_XL8(XINST_CREATE_store_1byte(
                       drcontext,
-                      OPND_CREATE_MEM8(regaddr, dr_page_size() /* enter redzone */),
-                      opnd_create_reg(reg_resize_to_opsz(scratch, OPSZ_1))),
+                      OPND_CREATE_MEM8(scratch_reg, dr_page_size() /* try enter redzone */),
+                      opnd_create_reg(reg_resize_to_opsz(scratch_reg2, OPSZ_1))),
                   instr_get_app_pc(where)));
+}
 
-    if (drreg_unreserve_register(drcontext, ilist, where, regaddr) != DRREG_SUCCESS ||
-        drreg_unreserve_register(drcontext, ilist, where, scratch) != DRREG_SUCCESS ||
-        drreg_unreserve_aflags(drcontext, ilist, where) != DRREG_SUCCESS)
-        DR_ASSERT(false);
+static bool
+instr_uses_mem(instr_t *where)
+{
+    int i;
+    for (i = 0; i < instr_num_srcs(where); i++) {
+        if (opnd_is_memory_reference(instr_get_src(where, i)))
+          return true;
+    }
+    for (i = 0; i < instr_num_dsts(where); i++) {
+        if (opnd_is_memory_reference(instr_get_dst(where, i)))
+            return true;
+    }
+
+    return false;
 }
 
 static dr_emit_flags_t
 event_app_instruction(void *drcontext, void *tag, instrlist_t *ilist, instr_t *where,
                       bool for_trace, bool translating, void *user_data)
 {
+    reg_id_t scratch_reg;
+    reg_id_t scratch_reg2;
+    instr_t *label = NULL;
+    bool uses_mem = instr_uses_mem(where);
+
+    if (uses_mem){
+
+    #ifdef X86
+        drvector_t allowed;
+        drreg_init_and_fill_vector(&allowed, false);
+        drreg_set_vector_entry(&allowed, DR_REG_XCX, true);
+    #endif
+
+        if (drreg_reserve_aflags(drcontext, ilist, where) != DRREG_SUCCESS ||
+            drreg_reserve_register(drcontext, ilist, where, IF_X86_ELSE(&allowed, NULL),
+                                   &scratch_reg2) != DRREG_SUCCESS ||
+            drreg_reserve_register(drcontext, ilist, where, NULL, &scratch_reg) !=
+                DRREG_SUCCESS) {
+            DR_ASSERT(false); /* Can't recover. */
+      }
+  #ifdef X86
+      drvector_delete(&allowed);
+  #endif
+
+   opnd_t flag_opnd = opnd_create_abs_addr(did_redzone_fault, OPSZ_1);
+   instr_t * check_instr = INSTR_CREATE_cmp(drcontext, flag_opnd, opnd_create_immed_int(0, OPSZ_1));
+   instrlist_meta_preinsert(ilist, where, check_instr);
+
+   label = INSTR_CREATE_label(drcontext);
+   instr_t * jmp_instr = INSTR_CREATE_jcc(drcontext, OP_jnz, opnd_create_instr(label));
+   instrlist_meta_preinsert(ilist, where, jmp_instr);
+  }
```

Comment:
Indentation of thie 7 or so lines above is off

Reply:
Done.